dart/lib/src/reference.dart

```diff
@@ -0,0 +1,416 @@
+    return null;
+  }
+
+  /// Returns [Uint8List] value or null otherwise.
+  Uint8List get blobValue {
+    if (_valueType == ValueType.Blob) {
+      return _buffer.buffer.asUint8List(_indirect, length);
+    }
+    return null;
+  }
+
+  /// Can be used with an [int] or a [String] value for key.
+  /// If the underlying value in FlexBuffer is a vector, then use [int] for access.
+  /// If the underlying value in FlexBuffer is a map, then use [String] for access.
+  /// Returns [Reference] value or null. Does not throw out of bounds exception.
+  Reference operator [](Object key) {
+    if (key is int && ValueTypeUtils.isAVector(_valueType)) {
+      var index = key;
+      if(index >= length) {
+        return null;
+      }
+      var elementOffset = _indirect + index * _byteWidth;
+      var flx = Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), 0);
+      flx._byteWidth = 1;
+      if (ValueTypeUtils.isTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.typedVectorElementType(_valueType);
+        return flx;
+      }
+      if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.fixedTypedVectorElementType(_valueType);
+        return flx;
+      }
+      var packedType = _buffer.getUint8(_indirect + length * _byteWidth + index);
+      return Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), packedType);
+    }
+    if (key is String && _valueType == ValueType.Map) {
+      var index = _keyIndex(key);
+      if (index != null) {
+        return _valueForIndex(index);
+      }
+    }
+    return null;
+  }
+
+  /// Get an iterable if the underlying flexBuffer value is a vector.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get vectorIterable {
+    if(isVector == false) {
+      throw Exception('Value is not a vector. It is: ${_valueType}');
+    }
+    return _VectorIterator(this);
+  }
+
+  /// Get an iterable for keys if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<String> get mapKeyIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapKeyIterator(this);
+  }
+
+  /// Get an iterable for values if the underlying flexBuffer value is a map.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get mapValueIterable {
+    if(isMap == false) {
+      throw Exception('Value is not a map. It is: ${_valueType}');
+    }
+    return _MapValueIterator(this);
+  }
+
+  /// Returns the length of the the underlying FlexBuffer value.
+  /// If the underlying value is [null] the length is 0.
+  /// If the underlying value is a number, or a bool, the length is 1.
+  /// If the underlying value is a vector, or map, the length reflects number of elements / element pairs.
+  /// If the values is a string or a blob, the length reflects a number of bytes the value occupies (strings are encoded in utf8 format).
```

Comment:
@aardappel - is this just how FlexBuffers work?

It seems surprising to me that `null` has a 0 length.

If so, we should probably document in here that this is the length of the value _excluding_ whatever byte(s) necessary to declare the type.

Reply:
This point has the same root as the previous point. Due to the fact that `Reference` is polymorphic and can represent any kind of value, the length is defined as it is. If the length of `null` should be `1` and not `0` is debatable :). The length method is however mainly used when the underlying value is a string, blob, vector, or a map.

Reply:
If that's how it's defined in other languages that's fine, but again I find it surprising that something taking space in my buffer would have a lengh of zero. It would be nice to document what this value means in terms of actual byte size in the buffer.